Glassfish 6.2.2 CDI deployment failure:null

## Environment Details
* GlassFish Version (and build number): 6.2.2
* JDK version: zulu 11.50.19, Spring Framework 1.3.5
* OS: Windows 10
* Database: Oracle 19c, derby 10

----------

## Problem Description
# "Describe the bug in detail highlighting current behavior vs expected behavior"
I have upgraded my application from java8 to java11. However I upgraded glassfish from v.4 to v.6.2.2 to support Java11. The application is built up successfully  however when I try to deploy it, it gives me an error. For the upgrade I have included different dependencies to cover some things that were missing from java 11. These were:

```
javax.annotation->javax.annotation-api@1.3.2
javax.xml.bind->jaxb-api@2.3.0
org.glassfish.jaxb->jaxb-runtime@2.3.0
com.sun.xml.bind->jaxb-core@2.3.0
com.sun.xml.bind->jaxb-impl@2.3.0
javax.activation->activation@1.1.1
```

# "Include stack traces or command outputs"

```
    [2021-10-19T21:18:15.879+0300] [glassfish 6.2] [SEVERE] [NCLS-CORE-00026] [jakarta.enterprise.system.core] [tid: _ThreadID=61 _ThreadName=admin-listener(3)] [timeMillis: 1634667495879] [levelValue: 1000] [[
      Exception during lifecycle processing
    org.glassfish.deployment.common.DeploymentException: CDI deployment failure:null
    	at org.glassfish.weld.WeldDeployer.event(WeldDeployer.java:211)
    	at org.glassfish.kernel.event.EventsImpl.send(EventsImpl.java:107)
    	at org.glassfish.internal.data.ApplicationInfo.load(ApplicationInfo.java:304)
    	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:473)
    	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:196)
    	at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:467)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:516)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:512)
    	at java.base/java.security.AccessController.doPrivileged(Native Method)
    	at java.base/javax.security.auth.Subject.doAs(Subject.java:361)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:511)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:542)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:534)
    	at java.base/java.security.AccessController.doPrivileged(Native Method)
    	at java.base/javax.security.auth.Subject.doAs(Subject.java:361)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:533)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1441)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1823)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1699)
    	at org.glassfish.admin.rest.resources.admin.CommandResource.executeCommand(CommandResource.java:359)
    	at org.glassfish.admin.rest.resources.admin.CommandResource.execCommandSimpInMultOut(CommandResource.java:202)
    	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
    	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
    	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
    	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
    	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
    	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:475)
    	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:397)
    	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
    	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
    	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
    	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
    	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
    	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
    	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
    	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
    	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
    	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
    	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:356)
    	at org.glassfish.admin.rest.adapter.JerseyContainerCommandService$3.service(JerseyContainerCommandService.java:150)
    	at org.glassfish.admin.rest.adapter.RestAdapter.service(RestAdapter.java:156)
    	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:440)
    	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:144)
    	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:174)
    	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:153)
    	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:196)
    	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:88)
    	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:246)
    	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:178)
    	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:118)
    	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:96)
    	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:51)
    	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:510)
    	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:82)
    	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:83)
    	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:34)
    	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:101)
    	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:535)
    	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:515)
    	at java.base/java.lang.Thread.run(Thread.java:829)
    Caused by: java.lang.UnsupportedOperationException
    	at java.base/java.util.AbstractList.add(AbstractList.java:153)
    	at java.base/java.util.AbstractList.add(AbstractList.java:111)
    	at org.glassfish.weld.DeploymentImpl.loadBeanDeploymentArchive(DeploymentImpl.java:416)
    	at org.jboss.weld.util.DeploymentStructures.getOrCreateBeanDeployment(DeploymentStructures.java:37)
    	at org.jboss.weld.bootstrap.ExtensionBeanDeployer.deployBean(ExtensionBeanDeployer.java:82)
    	at org.jboss.weld.bootstrap.ExtensionBeanDeployer.deployBeans(ExtensionBeanDeployer.java:75)
    	at org.jboss.weld.bootstrap.WeldStartup.startInitialization(WeldStartup.java:398)
    	at org.jboss.weld.bootstrap.WeldBootstrap.startInitialization(WeldBootstrap.java:79)
    	at org.glassfish.weld.WeldDeployer.event(WeldDeployer.java:198)
    	... 63 more
    ]]
    
    [2021-10-19T21:18:15.880+0300] [glassfish 6.2] [SEVERE] [] [jakarta.enterprise.system.core] [tid: _ThreadID=61 _ThreadName=admin-listener(3)] [timeMillis: 1634667495880] [levelValue: 1000] [[
      Exception while loading the app]]
    
    [2021-10-19T21:18:15.881+0300] [glassfish 6.2] [SEVERE] [AS-WEB-GLUE-00192] [jakarta.enterprise.web] [tid: _ThreadID=61 _ThreadName=admin-listener(3)] [timeMillis: 1634667495881] [levelValue: 1000] [[
      Undeployment failed for context /chesar3]]
    
    [2021-10-19T21:18:15.907+0300] [glassfish 6.2] [SEVERE] [] [jakarta.enterprise.system.core] [tid: _ThreadID=61 _ThreadName=admin-listener(3)] [timeMillis: 1634667495907] [levelValue: 1000] [[
      Exception while loading the app : CDI deployment failure:null
    java.lang.UnsupportedOperationException
    	at java.base/java.util.AbstractList.add(AbstractList.java:153)
    	at java.base/java.util.AbstractList.add(AbstractList.java:111)
    	at org.glassfish.weld.DeploymentImpl.loadBeanDeploymentArchive(DeploymentImpl.java:416)
    	at org.jboss.weld.util.DeploymentStructures.getOrCreateBeanDeployment(DeploymentStructures.java:37)
    	at org.jboss.weld.bootstrap.ExtensionBeanDeployer.deployBean(ExtensionBeanDeployer.java:82)
    	at org.jboss.weld.bootstrap.ExtensionBeanDeployer.deployBeans(ExtensionBeanDeployer.java:75)
    	at org.jboss.weld.bootstrap.WeldStartup.startInitialization(WeldStartup.java:398)
    	at org.jboss.weld.bootstrap.WeldBootstrap.startInitialization(WeldBootstrap.java:79)
    	at org.glassfish.weld.WeldDeployer.event(WeldDeployer.java:198)
    	at org.glassfish.kernel.event.EventsImpl.send(EventsImpl.java:107)
    	at org.glassfish.internal.data.ApplicationInfo.load(ApplicationInfo.java:304)
    	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:473)
    	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:196)
    	at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:467)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:516)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:512)
    	at java.base/java.security.AccessController.doPrivileged(Native Method)
    	at java.base/javax.security.auth.Subject.doAs(Subject.java:361)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:511)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:542)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:534)
    	at java.base/java.security.AccessController.doPrivileged(Native Method)
    	at java.base/javax.security.auth.Subject.doAs(Subject.java:361)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:533)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1441)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1823)
    	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1699)
    	at org.glassfish.admin.rest.resources.admin.CommandResource.executeCommand(CommandResource.java:359)
    	at org.glassfish.admin.rest.resources.admin.CommandResource.execCommandSimpInMultOut(CommandResource.java:202)
    	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
    	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
    	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
    	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
    	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
    	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:475)
    	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:397)
    	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
    	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
    	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
    	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
    	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
    	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
    	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
    	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
    	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
    	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
    	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:356)
    	at org.glassfish.admin.rest.adapter.JerseyContainerCommandService$3.service(JerseyContainerCommandService.java:150)
    	at org.glassfish.admin.rest.adapter.RestAdapter.service(RestAdapter.java:156)
    	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:440)
    	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:144)
    	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:174)
    	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:153)
    	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:196)
    	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:88)
    	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:246)
    	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:178)
    	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:118)
    	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:96)
    	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:51)
    	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:510)
    	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:82)
    	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:83)
    	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:34)
    	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:101)
    	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:535)
    	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:515)
    	at java.base/java.lang.Thread.run(Thread.java:829)
    ]]


```

## Actions Taken (didn't work)

Changes of bean.xml
```
<?xml version="1.0" encoding="UTF-8"?>
<beans xmlns="http://xmlns.jcp.org/xml/ns/javaee"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/beans_1_1.xsd"
       bean-discovery-mode="none">
<!--
This file is here for bean-discovery-mode="none", which turns off CDI 1.1's erroneous behaviour
of assuming that anything annotated with @Inject is a "managed bean" and must be instantiated
-->
</beans>
```

Deploying without CDI

`asadmin deploy --properties enable-implicit-cdi=false chesar3.war`

Deactivating CDI globally

`asadmin set configs.config.server-config.cdi-service.enable-implicit-cdi=false`



## Steps to reproduce
It's a whole application and I don't at which point there is the issue

## Impact of Issue
[//]: # "How has this issue affected you? What are you trying to accomplish?"
The application is not working

--- a/appserver/web/weld-integration/src/main/java/org/glassfish/weld/DeploymentImpl.java
+++ b/appserver/web/weld-integration/src/main/java/org/glassfish/weld/DeploymentImpl.java
@@ -134,7 +134,7 @@ public class DeploymentImpl implements CDI11Deployment {
             logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE, new Object[] { beanClass });
         }
 
-        List<BeanDeploymentArchive> beanDeploymentArchives = getBeanDeploymentArchives();
+        // Check if we have already created a bean archive for this bean class, and if so return it.
 
         for (BeanDeploymentArchive beanDeploymentArchive : beanDeploymentArchives) {
             if (logger.isLoggable(FINE)) {
@@ -197,29 +197,34 @@ public class DeploymentImpl implements CDI11Deployment {
                 new HashSet<>(),
                 context);
 
+        // If bean.xml explicitly says to ignore this archive, return
+        // without adding the bean archive
+
         BeansXml beansXml = newBeanDeploymentArchive.getBeansXml();
-        if (beansXml == null || !beansXml.getBeanDiscoveryMode().equals(NONE)) {
-            if (logger.isLoggable(FINE)) {
-                logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_ADD_NEW_BDA_TO_ROOTS, new Object[] {});
-            }
-
-            for (BeanDeploymentArchive beanDeploymentArchive : beanDeploymentArchives) {
-                beanDeploymentArchive.getBeanDeploymentArchives().add(newBeanDeploymentArchive);
-            }
-
-            if (logger.isLoggable(FINE)) {
-                logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_RETURNING_NEWLY_CREATED_BDA,
-                        new Object[] { beanClass, newBeanDeploymentArchive });
-            }
-
-            beanDeploymentArchives.add(newBeanDeploymentArchive);
-            idToBeanDeploymentArchive.put(newBeanDeploymentArchive.getId(), newBeanDeploymentArchive);
-            extensionBDAMap.put(beanClass.getClassLoader(), newBeanDeploymentArchive);
-
-            return newBeanDeploymentArchive;
+        if (beansXml != null && beansXml.getBeanDiscoveryMode().equals(NONE)) {
+            return null;
         }
 
-        return null;
+        // Add the new BeanDeploymentArchive to all root BeanDeploymentArchives of this deployment.
+
+        if (logger.isLoggable(FINE)) {
+            logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_ADD_NEW_BDA_TO_ROOTS, new Object[] {});
+        }
+
+        for (BeanDeploymentArchive beanDeploymentArchive : beanDeploymentArchives) {
+            beanDeploymentArchive.getBeanDeploymentArchives().add(newBeanDeploymentArchive);
+        }
+
+        if (logger.isLoggable(FINE)) {
+            logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_RETURNING_NEWLY_CREATED_BDA,
+                    new Object[] { beanClass, newBeanDeploymentArchive });
+        }
+
+        beanDeploymentArchives.add(newBeanDeploymentArchive);
+        idToBeanDeploymentArchive.put(newBeanDeploymentArchive.getId(), newBeanDeploymentArchive);
+        extensionBDAMap.put(beanClass.getClassLoader(), newBeanDeploymentArchive);
+
+        return newBeanDeploymentArchive;
     }
 
     @Override
@@ -329,7 +334,7 @@ public class DeploymentImpl implements CDI11Deployment {
     public void scanArchive(ReadableArchive archive, Collection<EjbDescriptor> ejbs, DeploymentContext context) {
         if (libJarRootBdas == null) {
             libJarRootBdas = scanForLibJars(archive, ejbs, context);
-            if (libJarRootBdas != null && libJarRootBdas.size() > 0) {
+            if (!isEmpty(libJarRootBdas)) {
                 return;
             }
         }

--- a/appserver/web/weld-integration/src/main/java/org/glassfish/weld/DeploymentImpl.java
+++ b/appserver/web/weld-integration/src/main/java/org/glassfish/weld/DeploymentImpl.java
@@ -16,24 +16,32 @@
 
 package org.glassfish.weld;
 
+import static com.sun.enterprise.util.Utility.isAnyEmpty;
+import static com.sun.enterprise.util.Utility.isAnyNull;
+import static com.sun.enterprise.util.Utility.isEmpty;
 import static java.util.Collections.emptyList;
 import static java.util.logging.Level.FINE;
+import static org.glassfish.cdi.CDILoggerInfo.CREATING_DEPLOYMENT_ARCHIVE;
+import static org.glassfish.cdi.CDILoggerInfo.EXCEPTION_SCANNING_JARS;
 import static org.glassfish.cdi.CDILoggerInfo.GET_BEAN_DEPLOYMENT_ARCHIVES;
 import static org.glassfish.cdi.CDILoggerInfo.LOAD_BEAN_DEPLOYMENT_ARCHIVE;
+import static org.glassfish.cdi.CDILoggerInfo.LOAD_BEAN_DEPLOYMENT_ARCHIVE_ADD_NEW_BDA_TO_ROOTS;
 import static org.glassfish.cdi.CDILoggerInfo.LOAD_BEAN_DEPLOYMENT_ARCHIVE_ADD_TO_EXISTING;
 import static org.glassfish.cdi.CDILoggerInfo.LOAD_BEAN_DEPLOYMENT_ARCHIVE_CHECKING;
 import static org.glassfish.cdi.CDILoggerInfo.LOAD_BEAN_DEPLOYMENT_ARCHIVE_CHECKING_SUBBDA;
 import static org.glassfish.cdi.CDILoggerInfo.LOAD_BEAN_DEPLOYMENT_ARCHIVE_CREATE_NEW_BDA;
 import static org.glassfish.cdi.CDILoggerInfo.LOAD_BEAN_DEPLOYMENT_ARCHIVE_RETURNING_NEWLY_CREATED_BDA;
+import static org.glassfish.deployment.common.InstalledLibrariesResolver.getInstalledLibraries;
 import static org.glassfish.weld.WeldDeployer.WELD_BOOTSTRAP;
 import static org.glassfish.weld.connector.WeldUtils.JAR_SUFFIX;
 import static org.glassfish.weld.connector.WeldUtils.META_INF_BEANS_XML;
 import static org.glassfish.weld.connector.WeldUtils.SEPARATOR_CHAR;
+import static org.glassfish.weld.connector.WeldUtils.isImplicitBeanArchive;
+import static org.jboss.weld.bootstrap.spi.BeanDiscoveryMode.NONE;
 
 import java.io.IOException;
 import java.net.URI;
 import java.net.URISyntaxException;
-import java.net.URL;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Enumeration;
@@ -42,7 +50,6 @@ import java.util.HashSet;
 import java.util.Iterator;
 import java.util.LinkedList;
 import java.util.List;
-import java.util.ListIterator;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.CopyOnWriteArrayList;
@@ -52,15 +59,12 @@ import org.glassfish.api.deployment.DeploymentContext;
 import org.glassfish.api.deployment.archive.ReadableArchive;
 import org.glassfish.cdi.CDILoggerInfo;
 import org.glassfish.deployment.common.DeploymentContextImpl;
-import org.glassfish.deployment.common.InstalledLibrariesResolver;
 import org.glassfish.javaee.core.deployment.ApplicationHolder;
-import org.glassfish.weld.connector.WeldUtils;
 import org.glassfish.weld.connector.WeldUtils.BDAType;
 import org.jboss.weld.bootstrap.WeldBootstrap;
 import org.jboss.weld.bootstrap.api.ServiceRegistry;
 import org.jboss.weld.bootstrap.api.helpers.SimpleServiceRegistry;
 import org.jboss.weld.bootstrap.spi.BeanDeploymentArchive;
-import org.jboss.weld.bootstrap.spi.BeanDiscoveryMode;
 import org.jboss.weld.bootstrap.spi.BeansXml;
 import org.jboss.weld.bootstrap.spi.CDI11Deployment;
 import org.jboss.weld.bootstrap.spi.Metadata;
@@ -84,13 +88,13 @@ public class DeploymentImpl implements CDI11Deployment {
     private List<BeanDeploymentArchive> beanDeploymentArchives;
     private DeploymentContext context;
 
-    // A convenience Map to get BDA for a given BDA ID
+    // A convenience Map to get a BeanDeploymentArchive for a given BeanDeploymentArchive ID
     private Map<String, BeanDeploymentArchive> idToBeanDeploymentArchive = new HashMap<>();
     private SimpleServiceRegistry simpleServiceRegistry = null;
 
     private Logger logger = CDILoggerInfo.getLogger();
 
-    // holds BDA's created for extensions
+    // Holds BeanDeploymentArchives created for extensions
     private Map<ClassLoader, BeanDeploymentArchive> extensionBDAMap = new HashMap<>();
 
     private Iterable<Metadata<Extension>> extensions;
@@ -104,11 +108,11 @@ public class DeploymentImpl implements CDI11Deployment {
      * Produce <code>BeanDeploymentArchive</code>s for this <code>Deployment</code> from information from the provided
      * <code>ReadableArchive</code>.
      */
-    public DeploymentImpl(ReadableArchive archive, Collection<EjbDescriptor> ejbs, DeploymentContext context,
-            ArchiveFactory archiveFactory) {
+    public DeploymentImpl(ReadableArchive archive, Collection<EjbDescriptor> ejbs, DeploymentContext context, ArchiveFactory archiveFactory) {
         if (logger.isLoggable(FINE)) {
-            logger.log(FINE, CDILoggerInfo.CREATING_DEPLOYMENT_ARCHIVE, new Object[] { archive.getName() });
+            logger.log(FINE, CREATING_DEPLOYMENT_ARCHIVE, new Object[] { archive.getName() });
         }
+
         this.archiveFactory = archiveFactory;
         this.beanDeploymentArchives = new ArrayList<>();
         this.context = context;
@@ -117,32 +121,205 @@ public class DeploymentImpl implements CDI11Deployment {
         // If we've produced BDA(s) from any /lib jars, <code>return</code> as
         // additional BDA(s) will be produced for any subarchives (war/jar).
         libJarRootBdas = scanForLibJars(archive, ejbs, context);
-        if (libJarRootBdas != null && libJarRootBdas.size() > 0) {
+        if (!isEmpty(libJarRootBdas)) {
             return;
         }
 
         createModuleBda(archive, ejbs, context);
     }
 
-    private void addBeanDeploymentArchives(RootBeanDeploymentArchive bda) {
-        BDAType moduleBDAType = bda.getModuleBDAType();
-        if (moduleBDAType.equals(BDAType.WAR)) {
-            if (warRootBdas == null) {
-                warRootBdas = new ArrayList<>();
-            }
-            warRootBdas.add(bda);
-        } else if (moduleBDAType.equals(BDAType.JAR)) {
-            if (ejbRootBdas == null) {
-                ejbRootBdas = new ArrayList<>();
-            }
-            ejbRootBdas.add(bda);
-        } else if (moduleBDAType.equals(BDAType.RAR)) {
-            if (rarRootBdas == null) {
-                rarRootBdas = new ArrayList<>();
-            }
-            rarRootBdas.add(bda);
+    @Override
+    public BeanDeploymentArchive loadBeanDeploymentArchive(Class<?> beanClass) {
+        if (logger.isLoggable(FINE)) {
+            logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE, new Object[] { beanClass });
         }
+
+        List<BeanDeploymentArchive> beanDeploymentArchives = getBeanDeploymentArchives();
+
+        for (BeanDeploymentArchive beanDeploymentArchive : beanDeploymentArchives) {
+            if (logger.isLoggable(FINE)) {
+                logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_CHECKING, new Object[] { beanClass, beanDeploymentArchive.getId() });
+            }
+
+            if (((BeanDeploymentArchiveImpl) beanDeploymentArchive).getModuleBeanClasses().contains(beanClass.getName())) {
+
+                // Don't stuff this Bean Class into the BeanDeploymentArchive's beanClasses,
+                // as Weld automatically add theses classes to the BeanDeploymentArchive's bean Classes
+                if (logger.isLoggable(FINE)) {
+                    logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_ADD_TO_EXISTING, new Object[] { beanClass.getName(), beanDeploymentArchive });
+                }
+
+                return beanDeploymentArchive;
+            }
+
+            // XXX: As of now, we handle one-level. Ideally, a bean deployment
+            // descriptor is a composite and we should be able to search the tree
+            // and get the right BDA for the beanClass
+            if (!beanDeploymentArchive.getBeanDeploymentArchives().isEmpty()) {
+                for (BeanDeploymentArchive subBeanDeploymentArchive : beanDeploymentArchive.getBeanDeploymentArchives()) {
+                    Collection<String> moduleBeanClassNames = ((BeanDeploymentArchiveImpl) subBeanDeploymentArchive).getModuleBeanClasses();
+                    if (logger.isLoggable(FINE)) {
+                        logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_CHECKING_SUBBDA,
+                                new Object[] { beanClass, subBeanDeploymentArchive.getId() });
+                    }
+
+                    if (moduleBeanClassNames.contains(beanClass.getName())) {
+
+                        // Don't stuff this Bean Class into the BeanDeploymentArchive's beanClasses,
+                        // as Weld automatically add theses classes to the BeanDeploymentArchive's bean Classes
+                        if (logger.isLoggable(FINE)) {
+                            logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_ADD_TO_EXISTING,
+                                    new Object[] { beanClass.getName(), subBeanDeploymentArchive });
+                        }
+
+                        return subBeanDeploymentArchive;
+                    }
+                }
+            }
+        }
+
+        BeanDeploymentArchive extensionBeanDeploymentArchive = extensionBDAMap.get(beanClass.getClassLoader());
+        if (extensionBeanDeploymentArchive != null) {
+            return extensionBeanDeploymentArchive;
+        }
+
+        // If the beanDeploymentArchive was not found for the Class, create one and add it
+
+        if (logger.isLoggable(FINE)) {
+            logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_CREATE_NEW_BDA, new Object[] { beanClass });
+        }
+
+        BeanDeploymentArchive newBeanDeploymentArchive =
+            new BeanDeploymentArchiveImpl(
+                beanClass.getName(),
+                new ArrayList<>(List.of(beanClass)),
+                new CopyOnWriteArrayList<>(),
+                new HashSet<>(),
+                context);
+
+        BeansXml beansXml = newBeanDeploymentArchive.getBeansXml();
+        if (beansXml == null || !beansXml.getBeanDiscoveryMode().equals(NONE)) {
+            if (logger.isLoggable(FINE)) {
+                logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_ADD_NEW_BDA_TO_ROOTS, new Object[] {});
+            }
+
+            for (BeanDeploymentArchive beanDeploymentArchive : beanDeploymentArchives) {
+                beanDeploymentArchive.getBeanDeploymentArchives().add(newBeanDeploymentArchive);
+            }
+
+            if (logger.isLoggable(FINE)) {
+                logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_RETURNING_NEWLY_CREATED_BDA,
+                        new Object[] { beanClass, newBeanDeploymentArchive });
+            }
+
+            beanDeploymentArchives.add(newBeanDeploymentArchive);
+            idToBeanDeploymentArchive.put(newBeanDeploymentArchive.getId(), newBeanDeploymentArchive);
+            extensionBDAMap.put(beanClass.getClassLoader(), newBeanDeploymentArchive);
+
+            return newBeanDeploymentArchive;
+        }
+
+        return null;
     }
+
+    @Override
+    public ServiceRegistry getServices() {
+        if (simpleServiceRegistry == null) {
+            simpleServiceRegistry = new SimpleServiceRegistry();
+        }
+
+        return simpleServiceRegistry;
+    }
+
+    @Override
+    public Iterable<Metadata<Extension>> getExtensions() {
+        if (extensions != null) {
+            return extensions;
+        }
+
+        List<Metadata<Extension>> extensionsList = new ArrayList<>();
+        for (BeanDeploymentArchive beanDeploymentArchive : getBeanDeploymentArchives()) {
+            if (!(beanDeploymentArchive instanceof RootBeanDeploymentArchive)) {
+                extensions =
+                    context.getTransientAppMetaData(WELD_BOOTSTRAP, WeldBootstrap.class)
+                           .loadExtensions(((BeanDeploymentArchiveImpl) beanDeploymentArchive).getModuleClassLoaderForBDA());
+
+                if (extensions != null) {
+                    for (Metadata<Extension> beanDeploymentArchiveExtension : extensions) {
+                        extensionsList.add(beanDeploymentArchiveExtension);
+                    }
+                }
+            }
+        }
+
+        return extensionsList;
+    }
+
+    @Override
+    public List<BeanDeploymentArchive> getBeanDeploymentArchives() {
+        if (logger.isLoggable(FINE)) {
+            logger.log(FINE, GET_BEAN_DEPLOYMENT_ARCHIVES, new Object[] { beanDeploymentArchives });
+        }
+
+        if (!beanDeploymentArchives.isEmpty()) {
+            return beanDeploymentArchives;
+        }
+
+        return emptyList();
+    }
+
+    /**
+     * Get a BeanDeploymentArchive for the specified beanClass
+     *
+     * @param beanClass The beanClass to get the BeanDeploymentArchive for.
+     *
+     * @return If the beanClass is in the archive represented by the BeanDeploymentArchive then return that BeanDeploymentArchive.
+     * Otherwise if the class loader of the beanClass matches the module class loader of any of the root BeanDeploymentArchives
+     * then return that root BeanDeploymentArchive.
+     * Otherwise return null.
+     */
+    @Override
+    public BeanDeploymentArchive getBeanDeploymentArchive(Class<?> beanClass) {
+        if (beanClass == null) {
+            return null;
+        }
+
+        for (BeanDeploymentArchive beanDeploymentArchive : beanDeploymentArchives) {
+            BeanDeploymentArchiveImpl beanDeploymentArchiveImpl = (BeanDeploymentArchiveImpl) beanDeploymentArchive;
+            if (beanDeploymentArchiveImpl.getBeanClassObjects().contains(beanClass)) {
+                return beanDeploymentArchive;
+            }
+        }
+
+        // Find a root BeanDeploymentArchive
+        ClassLoader classLoader = beanClass.getClassLoader();
+
+        RootBeanDeploymentArchive rootBda = findRootBda(classLoader, ejbRootBdas);
+        if (rootBda == null) {
+            rootBda = findRootBda(classLoader, warRootBdas);
+            if (rootBda == null) {
+                rootBda = findRootBda(classLoader, libJarRootBdas);
+                if (rootBda == null) {
+                    rootBda = findRootBda(classLoader, rarRootBdas);
+                }
+            }
+        }
+
+        return rootBda;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder valBuff = new StringBuilder();
+        for (BeanDeploymentArchive bda : getBeanDeploymentArchives()) {
+            valBuff.append(bda.toString());
+        }
+
+        return valBuff.toString();
+    }
+
+
+    // #### Public methods
 
     /**
      * Produce <code>BeanDeploymentArchive</code>s for this <code>Deployment</code> from information from the provided
@@ -226,10 +403,9 @@ public class DeploymentImpl implements CDI11Deployment {
         // If there are any application (/lib) jars, make them accessible
 
         if (warRootBdas != null) {
-            ListIterator<RootBeanDeploymentArchive> warIter = warRootBdas.listIterator();
             boolean modifiedArchive = false;
-            while (warIter.hasNext()) {
-                RootBeanDeploymentArchive warRootBda = warIter.next();
+            for (RootBeanDeploymentArchive warRootBda : warRootBdas) {
+
                 BeanDeploymentArchive warModuleBda = warRootBda.getModuleBda();
                 if (ejbRootBdas != null) {
                     for (RootBeanDeploymentArchive ejbRootBda : ejbRootBdas) {
@@ -298,6 +474,83 @@ public class DeploymentImpl implements CDI11Deployment {
         addDependentBdas();
     }
 
+    public void cleanup() {
+        if (ejbRootBdas != null) {
+            ejbRootBdas.clear();
+        }
+
+        if (warRootBdas != null) {
+            warRootBdas.clear();
+        }
+
+        if (libJarRootBdas != null) {
+            libJarRootBdas.clear();
+        }
+
+        if (rarRootBdas != null) {
+            rarRootBdas.clear();
+        }
+
+        if (idToBeanDeploymentArchive != null) {
+            idToBeanDeploymentArchive.clear();
+        }
+    }
+
+    public BeanDeploymentArchive getBeanDeploymentArchiveForArchive(String archiveId) {
+        return idToBeanDeploymentArchive.get(archiveId);
+    }
+
+    public Iterator<RootBeanDeploymentArchive> getLibJarRootBdas() {
+        if (libJarRootBdas == null) {
+            return null;
+        }
+
+        return libJarRootBdas.iterator();
+    }
+
+    public Iterator<RootBeanDeploymentArchive> getRarRootBdas() {
+        if (rarRootBdas == null) {
+            return null;
+        }
+
+        return rarRootBdas.iterator();
+    }
+
+    public Collection<EjbDescriptor> getDeployedEjbs() {
+        return deployedEjbs;
+    }
+
+    protected void addDeployedEjbs(Collection<EjbDescriptor> ejbs) {
+        if (ejbs != null) {
+            deployedEjbs.addAll(ejbs);
+        }
+    }
+
+
+
+    // #### Private methods
+
+
+    private void addBeanDeploymentArchives(RootBeanDeploymentArchive bda) {
+        BDAType moduleBDAType = bda.getModuleBDAType();
+        if (moduleBDAType.equals(BDAType.WAR)) {
+            if (warRootBdas == null) {
+                warRootBdas = new ArrayList<>();
+            }
+            warRootBdas.add(bda);
+        } else if (moduleBDAType.equals(BDAType.JAR)) {
+            if (ejbRootBdas == null) {
+                ejbRootBdas = new ArrayList<>();
+            }
+            ejbRootBdas.add(bda);
+        } else if (moduleBDAType.equals(BDAType.RAR)) {
+            if (rarRootBdas == null) {
+                rarRootBdas = new ArrayList<>();
+            }
+            rarRootBdas.add(bda);
+        }
+    }
+
     private void addDependentBdas() {
         Set<BeanDeploymentArchive> additionalBdas = new HashSet<>();
         for (BeanDeploymentArchive oneBda : beanDeploymentArchives) {
@@ -318,178 +571,6 @@ public class DeploymentImpl implements CDI11Deployment {
         }
     }
 
-    @Override
-    public List<BeanDeploymentArchive> getBeanDeploymentArchives() {
-        if (logger.isLoggable(FINE)) {
-            logger.log(FINE, GET_BEAN_DEPLOYMENT_ARCHIVES, new Object[] { beanDeploymentArchives });
-        }
-        if (!beanDeploymentArchives.isEmpty()) {
-            return beanDeploymentArchives;
-        }
-
-        return emptyList();
-    }
-
-    @Override
-    public BeanDeploymentArchive loadBeanDeploymentArchive(Class<?> beanClass) {
-        if (logger.isLoggable(FINE)) {
-            logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE, new Object[] { beanClass });
-        }
-        List<BeanDeploymentArchive> beanDeploymentArchives = getBeanDeploymentArchives();
-
-        ListIterator<BeanDeploymentArchive> lIter = beanDeploymentArchives.listIterator();
-        while (lIter.hasNext()) {
-            BeanDeploymentArchive beanDeploymentArchive = lIter.next();
-            if (logger.isLoggable(FINE)) {
-                logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_CHECKING, new Object[] { beanClass, beanDeploymentArchive.getId() });
-            }
-
-            if (((BeanDeploymentArchiveImpl) beanDeploymentArchive).getModuleBeanClasses().contains(beanClass.getName())) {
-
-                // Don't stuff this Bean Class into the BDA's beanClasses,
-                // as Weld automatically add theses classes to the BDA's bean Classes
-                if (logger.isLoggable(FINE)) {
-                    logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_ADD_TO_EXISTING, new Object[] { beanClass.getName(), beanDeploymentArchive });
-                }
-                return beanDeploymentArchive;
-            }
-
-            // XXX: As of now, we handle one-level. Ideally, a bean deployment
-            // descriptor is a composite and we should be able to search the tree
-            // and get the right BDA for the beanClass
-            if (beanDeploymentArchive.getBeanDeploymentArchives().size() > 0) {
-                for (BeanDeploymentArchive subBeanDeploymentArchive : beanDeploymentArchive.getBeanDeploymentArchives()) {
-                    Collection<String> moduleBeanClassNames = ((BeanDeploymentArchiveImpl) subBeanDeploymentArchive).getModuleBeanClasses();
-                    if (logger.isLoggable(FINE)) {
-                        logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_CHECKING_SUBBDA,
-                                new Object[] { beanClass, subBeanDeploymentArchive.getId() });
-                    }
-
-                    if (moduleBeanClassNames.contains(beanClass.getName())) {
-
-                        // Don't stuff this Bean Class into the BDA's beanClasses,
-                        // as Weld automatically add theses classes to the BDA's bean Classes
-                        if (logger.isLoggable(FINE)) {
-                            logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_ADD_TO_EXISTING,
-                                    new Object[] { beanClass.getName(), subBeanDeploymentArchive });
-                        }
-
-                        return subBeanDeploymentArchive;
-                    }
-                }
-            }
-        }
-
-        BeanDeploymentArchive extensionBeanDeploymentArchive = extensionBDAMap.get(beanClass.getClassLoader());
-        if (extensionBeanDeploymentArchive != null) {
-            return extensionBeanDeploymentArchive;
-        }
-
-        // If the beanDeploymentArchive was not found for the Class, create one and add it
-        if (logger.isLoggable(FINE)) {
-            logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_CREATE_NEW_BDA, new Object[] { beanClass });
-        }
-
-        List<Class<?>> beanClasses = new ArrayList<>();
-        List<URL> beanXMLUrls = new CopyOnWriteArrayList<>();
-        Set<EjbDescriptor> ejbs = new HashSet<>();
-        beanClasses.add(beanClass);
-
-        BeanDeploymentArchive newBeanDeploymentArchive = new BeanDeploymentArchiveImpl(beanClass.getName(), beanClasses, beanXMLUrls, ejbs, context);
-        BeansXml beansXml = newBeanDeploymentArchive.getBeansXml();
-        if (beansXml == null || !beansXml.getBeanDiscoveryMode().equals(BeanDiscoveryMode.NONE)) {
-            if (logger.isLoggable(FINE)) {
-                logger.log(FINE, CDILoggerInfo.LOAD_BEAN_DEPLOYMENT_ARCHIVE_ADD_NEW_BDA_TO_ROOTS, new Object[] {});
-            }
-
-            lIter = beanDeploymentArchives.listIterator();
-            while (lIter.hasNext()) {
-                BeanDeploymentArchive bda = lIter.next();
-                bda.getBeanDeploymentArchives().add(newBeanDeploymentArchive);
-            }
-
-            if (logger.isLoggable(FINE)) {
-                logger.log(FINE, LOAD_BEAN_DEPLOYMENT_ARCHIVE_RETURNING_NEWLY_CREATED_BDA,
-                        new Object[] { beanClass, newBeanDeploymentArchive });
-            }
-
-            beanDeploymentArchives.add(newBeanDeploymentArchive);
-            idToBeanDeploymentArchive.put(newBeanDeploymentArchive.getId(), newBeanDeploymentArchive);
-            extensionBDAMap.put(beanClass.getClassLoader(), newBeanDeploymentArchive);
-            return newBeanDeploymentArchive;
-        }
-
-        return null;
-    }
-
-    @Override
-    public ServiceRegistry getServices() {
-        if (simpleServiceRegistry == null) {
-            simpleServiceRegistry = new SimpleServiceRegistry();
-        }
-
-        return simpleServiceRegistry;
-    }
-
-    @Override
-    public Iterable<Metadata<Extension>> getExtensions() {
-        if (extensions != null) {
-            return extensions;
-        }
-
-        List<BeanDeploymentArchive> beanDeploymentArchives = getBeanDeploymentArchives();
-        List<Metadata<Extension>> extnList = new ArrayList<>();
-        for (BeanDeploymentArchive beanDeploymentArchive : beanDeploymentArchives) {
-            if (!(beanDeploymentArchive instanceof RootBeanDeploymentArchive)) {
-                ClassLoader moduleClassLoader = ((BeanDeploymentArchiveImpl) beanDeploymentArchive).getModuleClassLoaderForBDA();
-                extensions = context.getTransientAppMetaData(WELD_BOOTSTRAP, WeldBootstrap.class)
-                        .loadExtensions(moduleClassLoader);
-
-                if (extensions != null) {
-                    for (Metadata<Extension> beanDeploymentArchiveExtension : extensions) {
-                        extnList.add(beanDeploymentArchiveExtension);
-                    }
-                }
-            }
-        }
-        return extnList;
-    }
-
-    @Override
-    public String toString() {
-        StringBuffer valBuff = new StringBuffer();
-        List<BeanDeploymentArchive> beanDeploymentArchives = getBeanDeploymentArchives();
-        ListIterator<BeanDeploymentArchive> lIter = beanDeploymentArchives.listIterator();
-        while (lIter.hasNext()) {
-            BeanDeploymentArchive bda = lIter.next();
-            valBuff.append(bda.toString());
-        }
-        return valBuff.toString();
-    }
-
-    public BeanDeploymentArchive getBeanDeploymentArchiveForArchive(String archiveId) {
-        return idToBeanDeploymentArchive.get(archiveId);
-    }
-
-    public void cleanup() {
-        if (ejbRootBdas != null) {
-            ejbRootBdas.clear();
-        }
-        if (warRootBdas != null) {
-            warRootBdas.clear();
-        }
-        if (libJarRootBdas != null) {
-            libJarRootBdas.clear();
-        }
-
-        if (rarRootBdas != null) {
-            rarRootBdas.clear();
-        }
-
-        if (idToBeanDeploymentArchive != null) {
-            idToBeanDeploymentArchive.clear();
-        }
-    }
 
     // This method creates and returns a List of BeanDeploymentArchives for each
     // Weld enabled jar under /lib of an existing Archive.
@@ -498,22 +579,23 @@ public class DeploymentImpl implements CDI11Deployment {
         ApplicationHolder holder = context.getModuleMetaData(ApplicationHolder.class);
         if (holder != null && holder.app != null) {
             String libDir = holder.app.getLibraryDirectory();
-            if (libDir != null && !libDir.isEmpty()) {
+            if (!isEmpty(libDir)) {
                 Enumeration<String> entries = archive.entries(libDir);
                 while (entries.hasMoreElements()) {
                     String entryName = entries.nextElement();
-                    // if a jar is directly in lib dir and not WEB-INF/lib/foo/bar.jar
+
+                    // If a jar is directly in lib dir and not WEB-INF/lib/foo/bar.jar
                     if (entryName.endsWith(JAR_SUFFIX) && entryName.indexOf(SEPARATOR_CHAR, libDir.length() + 1) == -1) {
                         try {
                             ReadableArchive jarInLib = archive.getSubArchive(entryName);
-                            if (jarInLib.exists(META_INF_BEANS_XML) || WeldUtils.isImplicitBeanArchive(context, jarInLib)) {
+                            if (jarInLib.exists(META_INF_BEANS_XML) || isImplicitBeanArchive(context, jarInLib)) {
                                 if (libJars == null) {
                                     libJars = new ArrayList<>();
                                 }
                                 libJars.add(jarInLib);
                             }
                         } catch (IOException e) {
-                            logger.log(FINE, CDILoggerInfo.EXCEPTION_SCANNING_JARS, new Object[] { e });
+                            logger.log(FINE, EXCEPTION_SCANNING_JARS, new Object[] { e });
                         }
                     }
                 }
@@ -540,7 +622,7 @@ public class DeploymentImpl implements CDI11Deployment {
     private void createLibJarBda(RootBeanDeploymentArchive rootLibBda) {
         BeanDeploymentArchive libModuleBda = rootLibBda.getModuleBda();
         BeansXml moduleBeansXml = libModuleBda.getBeansXml();
-        if (moduleBeansXml == null || !moduleBeansXml.getBeanDiscoveryMode().equals(BeanDiscoveryMode.NONE)) {
+        if (moduleBeansXml == null || !moduleBeansXml.getBeanDiscoveryMode().equals(NONE)) {
             addBdaToDeploymentBdas(rootLibBda);
             addBdaToDeploymentBdas(libModuleBda);
             if (libJarRootBdas == null) {
@@ -563,120 +645,14 @@ public class DeploymentImpl implements CDI11Deployment {
         }
     }
 
-    private void addBdaToDeploymentBdas(BeanDeploymentArchive bda) {
-        if (!beanDeploymentArchives.contains(bda)) {
-            beanDeploymentArchives.add(bda);
-            idToBeanDeploymentArchive.put(bda.getId(), bda);
-        }
-    }
-
-    // These are application libraries that reside outside of the ear.  They are usually specified by entries
-    // in the manifest.
-    //
-    // To test this put a jar in domains/domain1/lib/applibs and in its manifest make sure it has something like:
-    //                           Extension-Name: com.acme.extlib
-    // In a war's manifest put in something like:
-    //                           Extension-List: MyExtLib
-    //                           MyExtLib-Extension-Name: com.acme.extlib
-    private void processBdasForAppLibs(ReadableArchive archive, DeploymentContext context) {
-        List<RootBeanDeploymentArchive> libBdas = new ArrayList<>();
-        try {
-            // each appLib in context.getAppLibs is a URI of the form "file:/glassfish/runtime/trunk/glassfish6/glassfish/domains/domain1/lib/applibs/mylib.jar"
-            List<URI> appLibs = context.getAppLibs();
-
-            Set<String> installedLibraries = InstalledLibrariesResolver.getInstalledLibraries(archive);
-            if (appLibs != null && appLibs.size() > 0 && installedLibraries != null && installedLibraries.size() > 0) {
-                for (URI oneAppLib : appLibs) {
-                    for (String oneInstalledLibrary : installedLibraries) {
-                        if (oneAppLib.getPath().endsWith(oneInstalledLibrary)) {
-                            ReadableArchive libArchive = null;
-                            try {
-                                libArchive = archiveFactory.openArchive(oneAppLib);
-                                if (libArchive.exists(WeldUtils.META_INF_BEANS_XML)) {
-                                    String bdaId = archive.getName() + "_" + libArchive.getName();
-                                    RootBeanDeploymentArchive rootBda = new RootBeanDeploymentArchive(libArchive,
-                                            emptyList(), context, bdaId);
-                                    libBdas.add(rootBda);
-                                }
-                            } finally {
-                                if (libArchive != null) {
-                                    try {
-                                        libArchive.close();
-                                    } catch (Exception ignore) {
-                                    }
-                                }
-                            }
-                            break;
-                        }
-                    }
-                }
-            }
-        } catch (URISyntaxException | IOException e) {
-            //todo: log error
-        }
-
-        for (RootBeanDeploymentArchive oneBda : libBdas) {
-            createLibJarBda(oneBda);
-        }
-    }
-
-    protected void addDeployedEjbs(Collection<EjbDescriptor> ejbs) {
-        if (ejbs != null) {
-            deployedEjbs.addAll(ejbs);
-        }
-    }
-
-    public Collection<EjbDescriptor> getDeployedEjbs() {
-        return deployedEjbs;
-    }
-
-    /**
-     * Get a bda for the specified beanClass
-     *
-     * @param beanClass The beanClass to get the bda for.
-     *
-     * @return If the beanClass is in the archive represented by the bda then return that bda. Otherwise if the class loader
-     * of the beanClass matches the module class loader of any of the root bdas then return that root bda. Otherwise return
-     * null.
-     */
-    @Override
-    public BeanDeploymentArchive getBeanDeploymentArchive(Class<?> beanClass) {
-        if (beanClass == null) {
-            return null;
-        }
-
-        for (BeanDeploymentArchive oneBda : beanDeploymentArchives) {
-            BeanDeploymentArchiveImpl beanDeploymentArchiveImpl = (BeanDeploymentArchiveImpl) oneBda;
-            if (beanDeploymentArchiveImpl.getBeanClassObjects().contains(beanClass)) {
-                return oneBda;
-            }
-        }
-
-        // find a root bda
-        ClassLoader classLoader = beanClass.getClassLoader();
-
-        RootBeanDeploymentArchive rootBda = findRootBda(classLoader, ejbRootBdas);
-        if (rootBda == null) {
-            rootBda = findRootBda(classLoader, warRootBdas);
-            if (rootBda == null) {
-                rootBda = findRootBda(classLoader, libJarRootBdas);
-                if (rootBda == null) {
-                    rootBda = findRootBda(classLoader, rarRootBdas);
-                }
-            }
-        }
-
-        return rootBda;
-    }
-
     private RootBeanDeploymentArchive findRootBda(ClassLoader classLoader, List<RootBeanDeploymentArchive> rootBdas) {
-        if (rootBdas == null || classLoader == null) {
+        if (isAnyNull(rootBdas, classLoader)) {
             return null;
         }
 
-        for (RootBeanDeploymentArchive oneRootBda : rootBdas) {
-            if (classLoader.equals(oneRootBda.getModuleClassLoaderForBDA())) {
-                return oneRootBda;
+        for (RootBeanDeploymentArchive rootBeanDeploymentArchive : rootBdas) {
+            if (classLoader.equals(rootBeanDeploymentArchive.getModuleClassLoaderForBDA())) {
+                return rootBeanDeploymentArchive;
             }
         }
 
@@ -688,7 +664,7 @@ public class DeploymentImpl implements CDI11Deployment {
 
         BeanDeploymentArchive moduleBda = rootBda.getModuleBda();
         BeansXml moduleBeansXml = moduleBda.getBeansXml();
-        if (moduleBeansXml == null || !moduleBeansXml.getBeanDiscoveryMode().equals(BeanDiscoveryMode.NONE)) {
+        if (moduleBeansXml == null || !moduleBeansXml.getBeanDiscoveryMode().equals(NONE)) {
             addBdaToDeploymentBdas(rootBda);
             addBdaToDeploymentBdas(moduleBda);
             addBeanDeploymentArchives(rootBda);
@@ -709,19 +685,64 @@ public class DeploymentImpl implements CDI11Deployment {
         processBdasForAppLibs(archive, context);
     }
 
-    public Iterator<RootBeanDeploymentArchive> getLibJarRootBdas() {
-        if (libJarRootBdas == null) {
-            return null;
+    private void addBdaToDeploymentBdas(BeanDeploymentArchive bda) {
+        if (!beanDeploymentArchives.contains(bda)) {
+            beanDeploymentArchives.add(bda);
+            idToBeanDeploymentArchive.put(bda.getId(), bda);
         }
-
-        return libJarRootBdas.iterator();
     }
 
-    public Iterator<RootBeanDeploymentArchive> getRarRootBdas() {
-        if (rarRootBdas == null) {
-            return null;
+    // These are application libraries that reside outside of the ear.  They are usually specified by entries
+    // in the manifest.
+    //
+    // To test this put a jar in domains/domain1/lib/applibs and in its manifest make sure it has something like:
+    //                           Extension-Name: com.acme.extlib
+    // In a war's manifest put in something like:
+    //                           Extension-List: MyExtLib
+    //                           MyExtLib-Extension-Name: com.acme.extlib
+    private void processBdasForAppLibs(ReadableArchive archive, DeploymentContext context) {
+        List<RootBeanDeploymentArchive> libBdas = new ArrayList<>();
+
+        try {
+            // Each appLib in context.getAppLibs is a URI of the form
+            // "file:/glassfish/runtime/trunk/glassfish6/glassfish/domains/domain1/lib/applibs/mylib.jar"
+            List<URI> appLibs = context.getAppLibs();
+            Set<String> installedLibraries = getInstalledLibraries(archive);
+
+            if (!isAnyEmpty(appLibs, installedLibraries)) {
+                for (URI appLib : appLibs) {
+                    for (String installedLibrary : installedLibraries) {
+                        if (appLib.getPath().endsWith(installedLibrary)) {
+                            ReadableArchive libArchive = null;
+                            try {
+                                libArchive = archiveFactory.openArchive(appLib);
+                                if (libArchive.exists(META_INF_BEANS_XML)) {
+                                    libBdas.add(new RootBeanDeploymentArchive(
+                                        libArchive,
+                                        emptyList(),
+                                        context,
+                                        archive.getName() + "_" + libArchive.getName()));
+                                }
+                            } finally {
+                                if (libArchive != null) {
+                                    try {
+                                        libArchive.close();
+                                    } catch (Exception ignore) {
+                                    }
+                                }
+                            }
+                            break;
+                        }
+                    }
+                }
+            }
+        } catch (URISyntaxException | IOException e) {
+            //todo: log error
         }
 
-        return rarRootBdas.iterator();
+        for (RootBeanDeploymentArchive libBeanDeploymentArchive : libBdas) {
+            createLibJarBda(libBeanDeploymentArchive);
+        }
     }
+
 }

